chipset_enable.c

```diff
@@ -1456,6 +1456,125 @@ static int enable_flash_sb600(const struct programmer_cfg *cfg, struct pci_dev *
 	return ret;
 }
 
+static int enable_flash_fch(const struct programmer_cfg *cfg, struct pci_dev *dev, const char *name)
+{
+	uint32_t prot, length;
+	uint8_t reg, spi_alt_cs;
+	int ret = 0;
+	uint8_t *fch_spi;
+	bool read_prot, write_prot;
+
+	uint32_t spi_base = pci_read_long(dev, 0xa0) & 0xffffffc0; /* bits 31:9 */
+	if (spi_base == 0) {
+		msg_perr("Error: Invalid SPI BASE\n");
+		return ERROR_FATAL;
+	}
+
+	fch_spi = rphysmap("FCH SPIBASE", spi_base, 256);
+	if (fch_spi == NULL) {
+		msg_perr("Error: Failed to map SPI BASE\n");
+		return ERROR_FATAL;
+	}
+
+	read_prot = write_prot = false;
+
+	spi_alt_cs = mmio_readb(fch_spi + 0x1d);
+
+	/* ROM Protect registers active? */
+	if (spi_alt_cs & 0x8) {
+		/* Clear ROM protect 0-3. */
+		for (reg = 0x50; reg < 0x60; reg += 4) {
+			prot = pci_read_long(dev, reg);
+			/* No protection flags for this region?*/
+			if ((prot & 0x600) == 0)
+				continue;
+
+			length = ((prot & 0xff) + 1) << ((prot & 0x100) ? 16 : 12);
+			msg_pdbg("Chipset %s%sprotected flash from 0x%08x to 0x%08x, unlocking...",
+				  (prot & 0x200) ? "read " : "",
+				  (prot & 0x400) ? "write " : "",
+				  (prot & 0xfffff000),
+				  (prot & 0xfffff000) + (length - 1));
+
+			/* ROM Protect registers locked? */
+			if (spi_alt_cs & 0x20) {
+				read_prot = read_prot || (prot & 0x200);
+				write_prot = write_prot || (prot & 0x400);
+				msg_pdbg("FAILED!\n");
+				ret = -1;
+				continue;
+			} else {
+				mmio_writeb(spi_alt_cs & 0xf7, fch_spi + 0x1d);
+				spi_alt_cs = mmio_readb(fch_spi + 0x1d);
+				if (spi_alt_cs & 0x8) {
+					msg_pdbg("FAILED!\n");
+					read_prot = read_prot || (prot & 0x200);
+					write_prot = write_prot || (prot & 0x400);
+					ret = -1;
+					continue;
+				}
+			}
+			msg_pdbg("done.\n");
+		}
+
+		if (ret != 0)
+			msg_pwarn("ROM Protect registers are active and locked\n");
+	}
+
+	if (write_prot) {
+		msg_pwarn("At least some flash regions are write protected. For write operations,\n"
+			  "you should use a flash layout and include only writable regions. See\n"
+			  "manpage for more details.\n");
+	}
+
+	if (read_prot) {
+		msg_pwarn("At least some flash regions are read protected. You have to use a flash\n"
+			  "layout and include only accessible regions. For write operations, you'll\n"
+			  "additionally need the --noverify-all switch. See manpage for more details.\n");
+	}
+
+	internal_buses_supported &= BUS_LPC | BUS_SPI;
+
+	ret = sb600_probe_spi(cfg, dev);
+
+	/* Read ROM strap override register. */
+	OUTB(0xca, 0xcd6);
+	reg = INB(0xcd7);
+	reg &= 0x0e;
+	msg_pdbg("ROM strap override is %sactive", (reg & 0x04) ? "" : "not ");
+	if (reg & 0x04) {
+		switch ((reg & 0x0c) >> 2) {
```

Comment:
```suggestion
		switch (reg & 0x03) {
```
At least according to BKDG I have.

Reply:
Yup, my bad.